lib/spack/spack/cmd/ci.py

```diff
@@ -241,8 +241,9 @@ def ci_reindex(args):
     ci_mirrors = yaml_root["mirrors"]
     mirror_urls = [url for url in ci_mirrors.values()]
     remote_mirror_url = mirror_urls[0]
+    mirror = spack.mirror.Mirror(remote_mirror_url)
```

Comment:
I don't think I understand the question.  But in case it helps: for pipelines, the first mirror in the `spack.yaml` is the special one we use to decide what needs to be rebuilt by comparing the hashes in the environment with binary packages in that mirror.

Reply:
Mirrors may have a different fetch / push url. That means in config it can look like this

```yaml
mirrors:
  name: <url>
```

or like this

```yaml
mirrors:
  name:
    fetch:
      url: http://example.com
    push:
      url: s3://bucket/x
```

So, it's wrong for multiple reasons: 

1. it cannot work with the second type of config
2. if you do spack.config.get(...) the yaml may go through transformations that you're missing when you directly look at the yaml.

Currently it feels very brittle. Wouldn't it be much better to just refer to mirrors by name? How much work is that to fix?

Reply:
IIUC, you're suggesting the ci module might have some documented naming convention for the special mirrors it uses, and then use those names instead of just bare urls or accessing `yaml_root["mirrors"][0]?  

I think that makes sense, and I don't think it's *that* much work.  But there are a few such mirrors used throughout the `ci` and `cmd/ci` modules.  Do you want me to make a PR to do that?

The special mirrors I can think of off the top of my head include 1) the main stack mirror you pointed to above, indexed directly from the `spack.yaml`, 2) the buildcache override mirror url we specify to customize where binaries should be pushed for pipelines in the same stack, but on different branches, 3) some older and maybe shouldn't be supported any more mirrors like the `artifacts_buildcache_mirror` for when we used to pass buildcaches from one stage to the next via artifacts.  Maybe there are some others I'm forgetting, but those first two are the main ones.

Reply:
Ok, I think it's a bit out of scope for this PR (also because I'm not very familiar with the magic mirrors). If you can do that, that'd be great 🙃 also that gives us hopefully an opportunity to switch to https:// for fetching instead of s3:// urls for gitlab ci, in my experience it has much lower latency.